Issue #150: test: ajoute des tests unitaires pour le loader

caarmen opened the pull request:
Issue: #150 

* Ajoute les dépendances de test dans `requirements-test.txt`.
* Définit un fichier `docker-compose.yml` pour une base de données PostgreSQL dédiée aux tests.
* Crée des fixtures pytest pour démarrer l'image Docker et attendre que la base de données soit accessible.
* Crée un test unitaire pour la fonction `test_load_file_to_pg`.
* Intègre l'exécution des tests unitaires dans le workflow GitHub Actions.

caarmen committed:
test: ajoute des tests unitaires pour le loader.

* Ajoute des dependences pour les tests dans `requirements-test.txt`
* Définit un fichier docker-compose.yml pour une BDD PostgreSQL pour les tests.
* Crée des fixtures pytest pour lancer l'image docker et attendre jusqu'à ce que la BDD soit joingable.
* Crée un test unitaire qui teste la fonction `test_load_file_to_pg`.

caarmen committed:
test : paramétrise le test `test_load_file_to_pg`

Rend la fonction de test paramétrable, permettant ainsi de vérifier plusieurs types de fichiers différents avec la même fonction.

caarmen committed:
test : ajoute un nouveau scénario pour `test_load_file_to_pg`

Ajoute un test pour un fichier utilisant `;` comme séparateur.

caarmen committed:
test : ajoute un script pour lancer les tests

* Ajoute la librairie `pytest-cov` pour générer un rapport de couverture de code.
* Ajoute un script de lancement des tests, avec génération de rapports :
  - `reports/junit.xml` : rapport XML des résultats des tests.
  - `reports/coverage.xml` : rapport XML de couverture de code.
  - `report/htmlcov` : rapport HTML de couverture de code.

caarmen committed:
test : ajoute un workflow GitHub Actions pour les tests

caarmen commented on tests/load/test_loaders.py:
```diff
@@ -0,0 +1,56 @@
+    Recharger le module, après l'exécution de notre fixture qui définit les variables
+    d'environnement pour la connexion à la BDD de test.
+    """
+    importlib.reload(loaders)
```
ℹ️ Réflexion

Des alternatives à cet `importlib.reload()`:
* Setter les variables d'envrionnement en shell avant de lancer pytest. Comme ça, même au chargement du module, le module trouvera les env vars pour la bdd de test.
  - Inconvenient : si on fait pas gaffe, on peut oublier cette étape et lancer les tests sur la bdd de la prod. Le test ajouté ici n'utilise pas les tables de la prod, donc c'est peut-être pas hyper grave. Mais je pense que si on peut s'assurer de ne pas utiliser la bdd de prod pour les tests, c'est peut-être mieux.
* Refactoriser le code pour injecter les paramètres de connexion à la base, et/ou éventuellement les charger de façon "lazy" depuis l'environnement. Ceci implique des modifs qui dépassent le scope d'une PR dont le but est juste d'ajotuer des tests.

caarmen replied:
Suite à une discussion avec @soufianekhoudmi, l'idée serait plutôt la 2ème option. 

On pourrait : 
* lire les variables d'environnement depuis le `main.py`
* ouvrir une connexion à la BDD
* passer l'objet de connexion aux fonctions de loader qui en ont besoin.

Micro bonus : ça pourrait être légèrement plus rapide, comme on ne ferait pas de reconnexion à chaque appel de fonction (mais ce n'est pas forcément l'objectif de la modification).

A traiter dans une PR séparée.

greg-ware replied:
J'ai refctorisé la connection à la DB, avec load optionel des parametres depuis un fichier d'environnement

caarmen commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,48 @@
+      with:
+        path: reports/coverage.xml
+        # augmenter ce seuil quand on ajoute des tests.
+        minimum_coverage: 35
```
couverture minimale : 35% - La barre est basse 😅 

Il faut fournir ce paramètre `minimum_coverage`. Je pense qu'il est pas utile de le setter à 80% tant qu'on a pas encore beaucoup de tests. Ca va faire un commentaire avec :x: dans chaque PR 😁

caarmen commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,48 @@
+    - name: Run tests
+      run: bash -x scripts/run_tests.sh
+
+    - name: Publish Test Report
```
ℹ️ Pour info

Cette étape publie un "check" sur la PR comme suit : 

![image](https://github.com/user-attachments/assets/c403b371-c984-4bec-8522-e9b2e6df682c)

caarmen replied:
Ceci feature de poster ce check "2 tests run, 2 passed, 0 skipped, 0 failed" a été supprimé pour l'instant, à cause des soucis de permissions dans les PR de fork.

caarmen commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,48 @@
+      if: always()
+      with:
+        report_paths: 'reports/junit.xml'
+    - name: Post coverage report
```
ℹ️ Pour info
Cette étape publie un commentaire sur la PR comme suit : 

![image](https://github.com/user-attachments/assets/3cbb643e-8f9c-494b-afd0-58f8f4de8214)

caarmen replied:
Ceci feature de poster commentaire de couverture a été supprimé pour l'instant, à cause des soucis de permissions dans les PR de fork.

caarmen commented on scripts/run_tests.sh:
```diff
@@ -0,0 +1,4 @@
+rm -rf reports
```
Serait-il utile de documenter l'existence de ce script ? Si oui, pensez-vous que ce serait plus pertinent dans le `README.md` ou dans le `CONTRIBUER.md` ? 🤔

caarmen replied:
Il est maintenant documenté dans `CONTRIBUER.md`.

caarmen commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,48 @@
+      uses: actions/upload-artifact@v4
+      with:
+        name: test-reports
+        path: reports
```
ℹ️ Pour info:
Voici un exemple du [rapport archivé](https://github.com/caarmen/make-open-data/actions/runs/11619840818/artifacts/2130632846) (depuis un workflow [run](https://github.com/caarmen/make-open-data/actions/runs/11619840818)).

caarmen commented on requirements-test.txt:
```diff
@@ -0,0 +1,4 @@
+-r requirements.txt
```
à mettre dans requirements.txt

caarmen commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,48 @@
+name: Tests
```
renommer à "UnitTests"

caarmen commented on .github/workflows/tests.yml:
```diff
@@ -0,0 +1,48 @@
+name: Tests
+
+on:
+  push:
```
on peut le lancer sur push et pull_request, n'importe quelle branche

caarmen commented:
Je mets la PR en draft, le temps de prendre en compte tous les feedbacks, et corriger le :x: sur la ci.

caarmen committed:
ci: Supprimer les étapes "Publish Test Report" et "Post coverage report".

Nous rencontrons des difficultés lorsque des PRs sont ouvertes depuis des forks.

Nous pouvons eventuellement remettre ces étapes dans un deuxième temps.

Au moins les tests sont exécutés, et les resultats sont visibles dans le console de github actions ainsi que dans les artifacts archivés (depuis la "Summary" du workflow run).

caarmen commented:
> Je mets la PR en draft, le temps de prendre en compte tous les feedbacks, et corriger le ❌ sur la ci.

La PR est re-ouverte.

Les modifs : 
* Mettre les dependences dans `requirements.txt` (et non `requirements-test.txt`).
* Utiliser "unit test" au lieu de juste "test" pour le nouveau workflow.
* Lancer les unit tests sur les branches et PRs sans filtre (pas que sur main par exemple).
* Supprimer des actions dans le workflow unit tests qui ne fonctionnent pas avec des PRs de fork.
* Documenter le script qui lance les tests, dans `CONTRIBUER.md`.
